File: python/src/basecamp/services/cards.py
"""Cards service with a presence-aware ``update``.

BC3 permits a card's JSON update params exactly as submitted (bc3#12521):
``kanban/cards_controller`` builds them from ``card_params`` alone, so a key the
body never mentions is never written. An omitted ``due_on`` therefore leaves the
card's due date **unchanged**, and only an explicit ``""`` or null clears it. The
``{ due_on: nil }.merge(card_params)`` default that used to make every sparse
PUT destructive survives for HTML and turbo_stream web forms alone, which this
SDK never speaks.

A sparse PUT is consequently the correct thing to write, so ``update`` sends
exactly the fields the caller addressed in a single request. The
read-modify-write that used to guard ``due_on`` is gone with the behaviour it
guarded: it now costs a round-trip to preserve a value nothing threatens, and
its GET-then-PUT window was a race a single request does not have.

A clear is spelled ``""`` rather than a JSON null — null would violate body
compaction (SPEC section 18), and Rails casts a blank date to nil regardless.

``update_verbatim`` remains as the unadorned name for the same single PUT.
"""

from __future__ import annotations

from typing import Any

from basecamp.generated.services.cards import AsyncCardsService as _GeneratedAsyncCardsService
from basecamp.generated.services.cards import CardsService as _GeneratedCardsService

_UPDATE_DOC = """Update a card without disturbing fields you did not mention.

``due_on`` is tri-state:

* ``None`` (unaddressed) — the key is omitted and the due date is left alone
* ``""`` — the due date is cleared, sent as an explicit empty string
* a date — the due date is set

Nothing is resent on your behalf: the request body carries only the arguments
you passed. Echoing a field back would be actively wrong for assignees — BC3
filters incoming IDs through ``reachable_people``, so replaying an id belonging
to someone who has since lost board access would silently unassign them.
"""


class CardsService(_GeneratedCardsService):
    """Sync cards service with a presence-aware ``update``."""

    def update(
        self,
        *,
        card_id: int,
        title: str | None = None,
        content: str | None = None,
        due_on: str | None = None,
        assignee_ids: list[int] | None = None,
    ) -> dict[str, Any]:
        return self.update_verbatim(
            card_id=card_id,
            title=title,
            content=content,
            due_on=due_on,
            assignee_ids=assignee_ids,
        )

    update.__doc__ = _UPDATE_DOC


class AsyncCardsService(_GeneratedAsyncCardsService):
    """Async cards service with a presence-aware ``update``."""

    async def update(
        self,
        *,
        card_id: int,
        title: str | None = None,
        content: str | None = None,
        due_on: str | None = None,
        assignee_ids: list[int] | None = None,
    ) -> dict[str, Any]:
        return await self.update_verbatim(
            card_id=card_id,
            title=title,
            content=content,
            due_on=due_on,
            assignee_ids=assignee_ids,
        )

    update.__doc__ = _UPDATE_DOC

File: rust/basecamp-sdk/src/services/cards.rs
//! Cards: the generated wire methods plus the merge-safe `update` of SPEC §5 "Merge-Safe
//! Write Surface (Cards)".
//!
//! `PUT /card_tables/cards/{cardId}` merges: BC3 builds the update from the JSON params
//! as they arrive, so a field the body omits is left alone and only a field that is
//! present changes. That makes a sparse PUT safe, and [`CardsService::update`] sends one
//! carrying exactly what the caller addressed — no read before the write, nothing echoed
//! back. Assignees especially: BC3 screens incoming ids through the board's reachable
//! people, so resending a list read from the card would silently unassign anyone who has
//! since lost access.
//!
//! Clearing the due date is something to say, not to leave unsaid. An absent `due_on`
//! means "leave unchanged", so the clear is spelled `"due_on": ""`, which BC3 casts to
//! nil; an explicit `null` is never sent (SPEC §18 body compaction).

pub use crate::generated::services::cards::{CardsService, ListCardsParams};

use crate::error::Error;
use crate::generated::routes;
use crate::generated::types::{Card, UpdateCardRequestContent};
use crate::services::DateChange;

/// The fields a merge-safe [`CardsService::update`] may set. A `None` member stays off
/// the wire and the server leaves it unchanged; a `Some` member is sent as given, so
/// `Some(String::new())` and `Some(vec![])` clear.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct UpdateCardRequest {
    /// The title.
    pub title: Option<String>,
    /// Rich text content (HTML).
    pub content: Option<String>,
    /// The due date: `None` leaves it alone, `Some(DateChange::Clear)` sends `""` to
    /// remove it, `Some(DateChange::On(date))` sets it.
    pub due_on: Option<DateChange>,
    /// The complete list of assigned person ids.
    pub assignee_ids: Option<Vec<i64>>,
}

impl UpdateCardRequest {
    fn body(&self) -> UpdateCardRequestContent {
        UpdateCardRequestContent {
            title: self.title.clone(),
            content: self.content.clone(),
            due_on: match self.due_on {
                Some(DateChange::On(date)) => Some(date),
                Some(DateChange::Clear) | None => None,
            },
            assignee_ids: self.assignee_ids.clone(),
        }
    }
}

impl CardsService<'_> {
    /// Updates a card, touching only the fields the caller named, in one PUT with no
    /// preceding read.
    ///
    /// A due-date clear is the one body the generated request type cannot spell — its
    /// `due_on` is a typed date, and `""` is not one — so that case takes the SPEC §18
    /// carve-out: the generated request is serialised, `"due_on": ""` is set on it, and
    /// the body goes out on the generated `UpdateCard` route, where hooks and retry still
    /// see the operation under its own identity. Every other request goes through
    /// [`CardsService::update_verbatim`] unchanged.
    pub async fn update(&self, card_id: i64, request: &UpdateCardRequest) -> Result<Card, Error> {
        let body = request.body();
        match request.due_on {
            Some(DateChange::Clear) => {
                let mut value = serde_json::to_value(&body).map_err(|error| {
                    Error::usage(format!("request body could not be serialized: {error}"))
                })?;
                if let serde_json::Value::Object(members) = &mut value {
                    members.insert(
                        "due_on".to_string(),
                        serde_json::Value::String(String::new()),
                    );
                }
                let mut operation = self.client().operation(&routes::UPDATE_CARD, &[&card_id]);
                operation.json(&value)?;
                self.client().send(operation).await
            }
            _ => self.update_verbatim(card_id, &body).await,
        }
    }
}
